docs/devel/api_changes.md

```diff
@@ -0,0 +1,231 @@
+
+## Changing versioned APIs
+
+For most changes, you will probably find it easiest to change the versioned
+APIs first.  This forces you to think about how to make your change in a
+compatible way.  Rather than doing each step in every version, it's usually
+easier to do each versioned API one at a time, or to do all of one version
+before starting "all the rest".
+
+### Edit types.go
+
+The struct definitions for each API are in `pkg/api/<version>/types.go`.  Edit
+those files to reflect the change you want to make.
+
+### Edit defaults.go
+
+If your change includes new fields for which you will need default values, you
+need to add cases to `pkg/api/<version>/defaults.go`.  Of course, since you
+have added code, you have to add a test: `pkg/api/<version>/defaults_test.go`.
+
+Don't forget to run the tests!
+
+### Edit conversion.go
+
+Given that you have not yet changed the internal API, this might feel premature
+- that's because it is.  You don't yet have anything to convert to or from.  We
+will revisit this in the "internal API" section.  If you're doing this all in a
+different order (i.e. you started with the internal API), then you should jump
+to that topic below.  In the very rare case that you are making an incompatible
+change you might or might not want to do this now, but you will have to do more
+later.  The files you want are `pkg/api/<version>/conversion.go` and
+`pkg/api/<version>/conversion_test.go`.
+
+## Changing the internal API
+
+Now it is time to change in the internal API so your versioned changes can be
+used.
+
+### Edit types.go
+
+Similar to the versioned APIs, the struct definitions for the inetrnal API are
+in `pkg/api/types.go`.  Edit those files to reflect the change you want to
+make.  Keep in mind that the internal API must be able to express *all* of the
+versioned APIs.
+
+## Edit validation.go
+
+Most changes made to the APIs need some form of input validation.  Validation
+is currently done on internal objects in `pkg/api/validation/validation.go`.
+This validation is the one of the first opportunities we have to make a great
+user experience - good error messages and thorough validation help ensure that
+users are giving you what you expect and, when they don't, that they know why
+and how to fix it.  Think hard about the contents of `string` fields, the bounds
+of `int` fields and the requiredness/optionalness of fields.
+
+Of course, code needs tests - `pkg/api/validation/validation_test.go`.
+
+## Edit version conversions
+
+At this point you have both the versioned API changes and the internal API
+changes done.  If there are any notable differences - field names, types,
+structural change in particular - you must add some logic to convert versioned
+APIs to and from internal.  This conversion process is a "star" with the
+internal API at the center - every versioned API must be able to convert to the
+internal API, and vice-versa, but versioned APIs do not convert to other
```

Comment:
This presumably means that the effort required to make an internal API change is proportional to the number of external (versioned) API versions. Do we only aim support some maximum number of external API versions (presumably 2-3) to put a bound on this effort?